src/msw/textctrl.cpp

```diff
@@ -2861,6 +2861,62 @@ bool wxTextCtrl::MSWSetCharFormat(const wxTextAttr& style, long start, long end)
             cf.dwEffects |= CFE_STRIKEOUT;
         }
     }
+    wxTextAttrUnderlineType underlineType = style.GetUnderlineType();
+    switch( underlineType )
+    {
+        case wxTEXT_ATTR_UNDERLINE_NONE:
+            cf.bUnderlineType = CFU_UNDERLINENONE;
+            break;
+		case wxTEXT_ATTR_UNDERLINE_SOLID:
+            cf.bUnderlineType = CFU_UNDERLINE;
+            break;
+		case wxTEXT_ATTR_UNDERLINE_DOUBLE:
+            cf.bUnderlineType = CFU_UNDERLINEDOUBLE;
+            break;
+		case wxTEXT_ATTR_UNDERLINE_WAVE:
+            cf.bUnderlineType = CFU_UNDERLINEWAVE;
+            break;
+    }
+#if _RICHEDIT_VER >= 0x0800
+// The colours are coming from https://docs.microsoft.com/en-us/windows/desktop/api/tom/nf-tom-itextdocument2-geteffectcolor.
+// Not all values from wxTheColourDatabase are supported as can be seen from the code
+// Those are commented out currently
+    BYTE colour;
+    wxColour col = style.GetUnderlineColour();
+    if( col == wxTheColourDatabase->Find( "BLACK" ) )
+        colour = 0x01;
+    else if( col == wxTheColourDatabase->Find( "BLUE" ) )
+        colour = 0x02;
+    else if( col == wxTheColourDatabase->Find( "CYAN" ) )
+        colour = 0x03;
+    else if( col == wxTheColourDatabase->Find( "GREEN" ) )
+        colour = 0x04;
+    else if( col == wxTheColourDatabase->Find( "MAGENTA" ) )
+        colour = 0x05;
+    else if( col == wxTheColourDatabase->Find( "RED" ) )
+        colour = 0x06;
+    else if( col == wxTheColourDatabase->Find( "YELLOW" ) )
+        colour = 0x07;
+    else if( col == wxTheColourDatabase->Find( "WHITE" ) )
+        colour = 0x08;
+//    else if( col == wxTheColourDatabase->Find( "NAVY" ) )
```

Comment:
@MaartenBent ,
Are you referring to the name or the colour value?

While name may exist, MSDN specifically refer the appropriate RGB value as only supported.
And I don't know enough about colour naming to add those RGB values.

Do you?

Reply:
Yes, I was looking at the name. 

Maybe it is better to compare RGB directly (`col == wxColour(0,0,0)` for black) and add the msdn link with supported colours to the documentation.

Also, I think you shouldn't use black as default, but `0 - Text color`. Maybe by setting the default colour value to `wxNullColour`?

Reply:
@MaartenBent ,
So how do you propose to call it?
Somewhere should be a definition for all those missing colours so that the call to `SetFontUnderline(TRUE, wxTEXT_ATTR_UNDERLINE_SOLID, wxTEAL );`will succeed (Just like in the sample). But TEAL in our colour database is not defined as what appers to be a TEAL in the link I provided.
Its possible it is related to another old ticket on trac.wxwidgets.org about the colour mismatch, which was fixed and later reverted. But I didn't look at that one close enough.

Reply:
I would not use the colour database and colour names, buy only compare rgb values. You might be able to find a match for teal, but what if a user uses `wxColour("LIGHT STEEL BLUE")` or `wxColour(0,4,60)`? I don't think finding the closest matching colour is an option.

I would just update the documentation like:
```
@param colour
    The colour of the line.
    Note: on Windows, only a limited number of colours is available. The supported RGB values are specified <a href="https://docs.microsoft.com/en-us/windows/desktop/api/tom/nf-tom-itextdocument2-geteffectcolor#parameters">here</a>.
```

Reply:
@MaartenBent ,
The ticket I referenced is 4293.
It is referred to wxPython, but it is the same idea - different standard uses different values.

To answer your question - the function will just fal, because according to MSDN only those values are supported. But I still don't see how comparing RGB values will allow me to call `SetFontUnderline(TRUE, wxTEXT_ATTR_UNDERLINE_SOLID, wxTEAL );` and compare RGB values internally if the wxTEAL will not match the wx internal colour DB.

Just look at [this](https://stackoverflow.com/questions/1756263/how-to-change-underlining-color-in-a-rich-edit-control-win32-c).

And I did plan to make a note in the docs about this limitation.

Reply:
But there is no `wxTEAL` in the wxWidgets API. If you add it to the colour database (`{wxT("TEAL"), 0, 128, 128}`) then comparing RGB should just work.

Look at the sample, it uses `*wxRED`
```c++
SetFontUnderline( TRUE, wxTEXT_ATTR_UNDERLINE_WAVE, *wxRED )
```

When I modify the code in `MSWSetCharFormat` to this it still works:
```c++
    else if ( col == wxColour( 255, 0 , 0 ) )
        colour = 0x06;
```

Reply:
Well, wxRED works because its a stock color.
But if I add the new TEAL to the wxColourTable (which is basic C-array), it will break either this algorythm or the old code coloring scheme. Because now, wxColourDatabase will contain 2 values for the same color and the first one will be picked up (from wx source code):

```
    static const struct wxColourDesc
    {
        const wxChar *name;
        unsigned char r,g,b;
    }
    wxColourTable[] =

```

Or you meant something else?

Reply:
Why 2? There is no RGB(0, 128, 128) in the colour db.
And I prefer to not use names, but RGB directly.  If only to prevent wrong colour definitions like in 4293. Note that MSDN doesn't specific names either, only the RGB Value.

Reply:
@MaartenBent ,
Because if I add the TEAL to the wxColourDatabase array it will break existing code.
I looked at the wxColourDatabase::Find() implementation. It uses hash map to search the colour by name. And so a lot of the existing code will break because of different colour scheme.

I understand you like to write RGB(), but a lot of people that use wxWidgets will just write wxRED and forget about it.
Unfortunately if someone will prefer to use TEAL or some other color name that is already there and have "wrong" RGB value...

We can ask to use RGB values in this case, but wxColour and the colour database exists for a reason.